### ISSUE TITLE
NzNoise refresh

### ISSUE THREAD

**Issue by Overdrivr** [DigitalPulseSoftware/NazaraEngine#49]
So what is the strategy to port changes from [ZNoise][https://github.com/Overdrivr/ZNoise] ?

I am not really in favor to make a manual porting, because that would mean two repositories to maintain for the same base code. But I can do it, it's not that many file.

Do you see any other option ?

Also, this leads me to this next point. In a more general approach, this could be a good exercise on how to have Nazara modules that are not necessarily a part of the core distribution.

Maybe to push further the idea, it could very cool to be able to install non-core modules simply from a bash command similar to (or even just that) ```git clone``` and have the build system be fine with it.

What are your thoughts about these two points ?

**Comment by DPSLynix** [comment]
Hi!

Sorry to answer now, I know it's kinda late for better late than never!

It would be great to see an update for this module, I can make the port myself if you wish not to.

Your idea about including external modules is interesting, but I have no idea right now on how it could be done, do you have some?

**Comment by Overdrivr** [comment]
Well, I would make the changes myself, but I don't know if I will find the time for now.

In the end, what sort of changes would be needed to make the port ? Using the `nz::` namespace, do you see anything else ?

By the way, I have made some improvements over the exisiting interface, specially a clearer separation between the noise algorithm and the way several noises are mixed. 

This reduced a lot the amount of classes/code thanks to std::initializer_list which allows me to write a single mixer class for any number of dimensions (while before they had to be written individually for each class).

**Comment by DPSLynix** [comment]
There's not much that should change hopefully.

Talking about your interface, I think using std::initializer_list is a mistake, you're using it as a fixed-size array except for the Get method which should be overloaded.

I think your derived classes should only have an overloaded Get method, which would implement the _2D, _3D, etc. stuff depending on the number of arguments, I see no benefit using an std::initializer_list here.

**Comment by Overdrivr** [comment]
I can't recall exactly but I feel I tried overloading Get initially, but something made me move away from it. Can't remember for sure.
initializer_list allows me to support arbitrary number of dimensions for mixer classes with the same code.
Is there a big performance hit to use initializer_list ?

**Comment by DPSLynix** [comment]
initializer_list is as light as passing arguments, but I can see in your
code that you're dispatching the initializer_list arguments depending on
how many floats are contained inside of it, that could be done by the
compiler by overloading.

If you wish, I can make the changes myself on your repository and submit a
pull request, you'll then be free to accept/reject it :)

2016-03-08 15:15 GMT+01:00 Rémi Bèges <notifications@github.com>:

**Comment by Overdrivr** [comment]
So inside NoiseBase the following would be declared
```
virtual float Get(float x, float y, float scale) const;
virtual float Get(float x, float y, float z, float scale) const;
virtual float Get(float float x, float y, float z, float scale) const;
```

The only downside is for Mixer, where the [following method](https://github.com/Overdrivr/ZNoise/blob/master/src/znoise/cpp/FBM.cpp) will require to be implemented for each dimension. Or maybe do you see a way around that ?

**Comment by DPSLynix** [comment]
That's the spirit.

About Mixer::Get, I missed that one, but I still think your interface would
be way simplified by using overloads,

It would be possible to use tricks to simplify that one, but it seems it
could be declined in three versions without repeating much code.

2016-03-08 15:38 GMT+01:00 Rémi Bèges <notifications@github.com>:

**Comment by DPSLynix** [comment]
Do you allow me to take some part of your code into Nazara in my way? That could eventually give you some idea for your own repo.

**Comment by Overdrivr** [comment]
Yes sure, it's always interesting to get some feedback from you. I am way too caught up on other things these sorry I cannot help out, but I'll follow closely

**Comment by DPSLynix** [comment]
Hi,

I imported your repo as the new Noise Module: 3fc9d758bb12a8ac801d3dba6c2b8c40276aea8e

I also spent some times refactoring your code, I only did a few of the things I want to do, but I think it allows you to see what I propose: 6a5a9a3993b7e2fc117c70647984ece702532da6

Tell me what you think about it!

**Comment by DPSLynix** [comment]
@Overdrivr Can I get an answer to this? Thanks.

**Comment by Overdrivr** [comment]
Hi Lynix, sorry about the delay for responding.

Your changes look really good, it's always interesting to read code written by you.

I see you went back to the interface I initially employed, which sacrificies in some way code amount for more clarity (the mixers need to be implemented for each dimension although they do the exact same thing).

I cannot see a better interface that would allow that clarity without repeating the code for the mixers, so I do agree this is an improvement.

**Comment by SirLynix** [comment]
The Noise module has been removed from Nazara-Next from now, but may be added back again in the future.
